File: api-platform/src/models/notification.py
"""Notification model - 通知模型"""

import uuid
from datetime import datetime
from typing import Optional, List

from sqlalchemy import Column, String, Boolean, Integer, DateTime, Text, ForeignKey
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship

from src.config.database import Base


class Notification(Base):
    """Notification model - 通知表"""

    __tablename__ = "notifications"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    # 接收通知的用户ID
    user_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    
    # 通知类型：system（系统通知）、billing（账单通知）、api（API通知）、security（安全通知）
    notification_type = Column(String(30), nullable=False, default="system")
    
    # 通知标题
    title = Column(String(200), nullable=False)
    
    # 通知内容
    content = Column(Text, nullable=False)
    
    # 额外数据（JSON格式，如跳转链接、关联ID等）
    extra_data = Column(JSONB, default=dict)
    
    # 通知状态：unread（未读）、read（已读）、deleted（已删除）
    status = Column(String(20), nullable=False, default="unread")
    
    # 是否已删除
    is_deleted = Column(Boolean, default=False)
    
    # 优先级：low（低）、normal（普通）、high（高）、urgent（紧急）
    priority = Column(String(20), nullable=False, default="normal")
    
    # 过期时间（可选，过期后自动标记为已过期）
    expire_at = Column(DateTime, nullable=True)
    
    # 审计字段
    created_at = Column(DateTime, default=datetime.utcnow)
    read_at = Column(DateTime, nullable=True)  # 阅读时间

    def __repr__(self):
        return f"<Notification {self.title}>"


class NotificationPreference(Base):
    """Notification preference model - 用户通知偏好设置"""

    __tablename__ = "notification_preferences"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), nullable=False, unique=True, index=True)

    # 是否启用邮件通知
    email_enabled = Column(Boolean, default=True)
    
    # 是否启用站内通知
    in_app_enabled = Column(Boolean, default=True)
    
    # 是否启用推送通知
    push_enabled = Column(Boolean, default=False)
    
    # 通知类型偏好
    preferences = Column(JSONB, default={
        "system": True,
        "billing": True,
        "api": True,
        "security": True,
    })

    # 审计字段
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def __repr__(self):
        return f"<NotificationPreference user_id={self.user_id}>"

File: api-platform/web/src/api/notification.ts
/**
 * 通知 API 服务
 */

import { api } from './client'

// 通知类型
export type NotificationType = 'system' | 'billing' | 'api' | 'security'

// 通知优先级
export type NotificationPriority = 'low' | 'normal' | 'high' | 'urgent'

// 通知状态
export type NotificationStatus = 'unread' | 'read' | 'deleted'

// 通知项接口
export interface Notification {
  id: string
  title: string
  content: string
  notification_type: NotificationType
  priority: NotificationPriority
  status: NotificationStatus
  extra_data: Record<string, any>
  created_at: string
  read_at: string | null
}

// 通知列表响应
export interface NotificationListResponse {
  total: number
  page: number
  page_size: number
  notifications: Notification[]
}

// 未读数量响应
export interface UnreadCountResponse {
  unread_count: number
}

// 通知偏好设置
export interface NotificationPreference {
  email_enabled: boolean
  in_app_enabled: boolean
  push_enabled: boolean
  preferences: {
    system: boolean
    billing: boolean
    api: boolean
    security: boolean
  }
}

// 通知 API
export const notificationApi = {
  /**
   * 获取通知列表
   */
  getList: (params?: {
    status?: string
    notification_type?: string
    page?: number
    page_size?: number
  }) => {
    return api.get<NotificationListResponse>('/notifications', params)
  },

  /**
   * 获取未读通知数量
   */
  getUnreadCount: () => {
    return api.get<UnreadCountResponse>('/notifications/unread-count')
  },

  /**
   * 获取最近未读通知（用于下拉面板）
   */
  getRecent: (limit: number = 5) => {
    return api.get<Notification[]>('/notifications/recent', { limit })
  },

  /**
   * 标记单条通知为已读
   */
  markAsRead: (notificationId: string) => {
    return api.post(`/notifications/${notificationId}/read`)
  },

  /**
   * 标记所有通知为已读
   */
  markAllAsRead: () => {
    return api.post('/notifications/read-all')
  },

  /**
   * 删除通知
   */
  delete: (notificationId: string) => {
    return api.delete(`/notifications/${notificationId}`)
  },

  /**
   * 删除所有已读通知
   */
  deleteAllRead: () => {
    return api.delete('/notifications/read/delete-all')
  },

  /**
   * 获取通知偏好设置
   */
  getPreference: () => {
    return api.get<NotificationPreference>('/notifications/preferences')
  },

  /**
   * 更新通知偏好设置
   */
  updatePreference: (data: Partial<NotificationPreference>) => {
    return api.put<NotificationPreference>('/notifications/preferences', data)
  },
}

export default notificationApi
